proposals/2024-11-25_changing_wal_format.md

```diff
@@ -0,0 +1,260 @@
+ 
+In other words, we propose to add a TSDB flag `--storage.tsdb.stateful.write-wal-version=<version>` that tells Prometheus to use a particular WAL format for both WAL and WBL. This kind of flag will change its default to a new version ONLY when (at least) one previous Prometheus version can read that version (while writing the old one). The initial version would be `v1`.
+
+There are two reasons for this flag:
+
+* Allows users to get the new features sooner and skip the safety mechanism.
+* It simplifies the process as the flag default mechanism guides users and devs in the rollout and revert procedures e.g:
+  * when we switch to writing v2, it's clearly visible in a flag default value. 
+  * when we at some point remove support of WAL v1, it's clear when it happens (v1 flag value is removed).
+* It allows users to set new version to write old format if needed (de-risking further).
+* Gives devs quicker feedback, makes testing easier, and motivates further contributions.
+
+We propose to document that behaviour in:
+
+* dedicated guide for users
+* flag help
+* tsdb/docs for devs
+
+To achieve WAL versioning we also propose to start versioning the WAL, as [a whole format](https://github.com/prometheus/prometheus/blob/main/tsdb/docs/format/wal.md). This is necessary to communicate a breaking change and to tell Prometheus what WAL format to write in.
+
+We propose the addition of a `meta.json` file in the wal directory, similar to [block meta.json](https://github.com/prometheus/prometheus/blob/5e124cf4f2b9467e4ae1c679840005e727efd599/tsdb/block.go#L171), with Version field set to `1` for the current format and `2` for new changes e.g. when we start to write [new records](https://github.com/prometheus/prometheus/pull/15467/files). No `meta.json` is equivalent to `{"version":1} `meta.json` file.
+
+We propose to also store the new WALs in separate directories e.g. `wal.v2`. Thanks to that the rewrite from one version to another is **eventual** and can be done segment by segment without any forcible rewrite. The WAL will be rewritten within the next 2h of a normal operation. The additional advantage of this way of versioning is that it's clear when your WAL fully migrated to a certain version.
+
+Finally, we propose that all Prometheus releases will contain the following table:
+
+| Data        | Supported | Writes |
+|-------------|-----------|--------|
+| WAL         | v1, v2    | v1     |
+| Block index | v2        | v2     |
+
+```
+The last revertable Prometheus version: v3.0 
+```
+
+See [alternatives](#alternatives) for other ideas.
+
+### Two-Fold Migration Strategy
+
+Given the following example:
+
+![twofold.png](../assets/2024-11-25_changing_wal_format/twofold.png)
+
+1. We release Prometheus X+1 version that supports both Y and Y+1 data but still writes Y.
+2. We release Prometheus X+2 version that supports both Y and Y+1 data, but now it writes new data as Y.
```

Comment:
```
2. We release Prometheus X+2 version that supports both Y and Y+1 data, but now it writes new data as Y.
```
For clarity: does this mean that the new record type (Y+1) will be re-named to Y?

Reply:
Ups should be ` We release Prometheus X+2 version that supports both Y and Y+1 data, but now it writes new data as Y+1.`

Thanks for catching!